### ISSUE TITLE
IMAP Client does not seem to handle unsolicited server responses

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10722]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| cablehead reported|
|-|-|
|Trac ID|trac#1105|
|Type|defect|
|Created|2005-07-08 17:48:07Z|
|Branch|https://github.com/twisted/twisted/tree/imap4-unsolicited-responses-1105|

Attachments:

* [imaptest.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7cb/7cb3f42379df52fb7c22a44c27399c3177bd99d2/21ba1c6e52a72c54a5721161686d5d49116eb2a8.patch) (3071 bytes) - added by cablehead on 2005-08-08 15:36:18Z - 
* [imap4_flags.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7cb/7cb3f42379df52fb7c22a44c27399c3177bd99d2/28520af5d163d36d40a37b86211061578dd98a0a.py) (1014 bytes) - added by therve on 2006-02-25 15:24:29Z - Patch for unsolicited flags response on imap4
* [test_imap_flags.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7cb/7cb3f42379df52fb7c22a44c27399c3177bd99d2/40295235e9b0ec5f1d741efee34d752023eefb81.py) (2397 bytes) - added by therve on 2006-02-25 15:25:14Z - Test imap unsolocited flags messages
* [twistedmail-imap-flags-anywhere-test.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7cb/7cb3f42379df52fb7c22a44c27399c3177bd99d2/df556d930ca5e35d24dc9b357202e6bfc9f0c56e.diff) (4474 bytes) - added by jhohm on 2007-05-30 18:19:12Z - Test FLAGS anywhere inside FETCH response
* [twistedmail-imap-flags-anywhere.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7cb/7cb3f42379df52fb7c22a44c27399c3177bd99d2/ec8ae39406c4f0da8285167fccfc1cc7c0e1c6fb.diff) (2262 bytes) - added by jhohm on 2007-05-30 18:19:53Z - Patch to handle FLAGS anywhere inside FETCH response

<details><summary>Searchable metadata</summary>

```
trac-id__1105 1105
type__defect defect
reporter__cablehead cablehead
priority__highest highest
milestone__ 
branch__branches_imap4_unsolicited_responses_1105 branches/imap4-unsolicited-responses-1105
branch_author__jhohm__exarkun jhohm, exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1120844887000000 1120844887000000
changetime__1237809763000000 1237809763000000
version__ 
owner__ 
cc__exarkun cc__cablehead
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|cablehead commented|
|-|-|

```
#!html
<pre>
http://twistedmatrix.com/pipermail/twisted-python/2005-July/010933.html
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|cablehead commented|
|-|-|

```
#!html
<pre>
stab at a test case for this issue
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve commented|
|-|-|

OK I managed to make something work. The problem relies in the Command, because fetch command doesn't check the data it received : if your fetch a SUBJECT, you should raise an error if your receive a BODY, or add the result in the unuse list (to handle FLAGS messages).

Attached a quick and dirty patch that handles the problem for FETCH that raises FLAGS messages. Also patch the test for not using defferedResult.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Test looks really nice, and the patch to imap4.py 'almost' looks cool.  Help me refresh my memory though... It's possible for responses other than FLAGS to come back unsolicited, isn't it?  So to be really correct, the patch needs to account for any response that wasn't explicitly requested?  Or am I imagining things?

**Comment by twisted-trac** [comment]
|<img alt="cablehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|cablehead commented|
|-|-|

Well, the rfc reads:

A client MUST be prepared to accept any server response at all times.

In practice though, this has only been an issue for me with FLAGS.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [16253]) Apply patch from cablehead/therve - adds support for recognizing unsolicited FLAGS responses to IMAP4Client; Fixes [#10722](https://github.com/twisted/twisted/issues/10722)

**Comment by twisted-trac** [comment]
|<img alt="jhohm's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|jhohm set status to `reopened`|
|-|-|

The already-applied patch is not sufficient.  It only recognizes unsolicited FETCH responses which contain only FLAGS, like this (for explanatory purposes the entire RFC822 body of the email is just 0123456789 followed by a CRLF):

```
* 1 FETCH (UID 9 RFC822 {12}
0123456789
)
* 1 FETCH (FLAGS (\Seen))
```

Cyrus IMAPD (version 2.2.13) sends FLAGS updates before the requested parts inside a single FETCH response, like this:

```
* 1 FETCH (FLAGS (\Seen) UID 9 RFC822 {12}
0123456789
)
```

Microsoft Exchange Server 2003 (version 6.5.7638.1) sends FLAGS updates after the requested parts inside a single FETCH response, like this:

```
* 1 FETCH (RFC822 {12}
0123456789
 UID 9 FLAGS (\Seen))
```

With the already-applied patch, such responses cause message data to be dropped from the fetch results, or cause an IllegalServerResponse exception.

I have attached a patch which add tests for FLAGS anywhere inside FETCH responses; I have attached a patch which removes the partial fix to Command from the applied patch and fixes the fetch callbacks to handle FLAGS anywhere inside FETCH responses; and I am reopening this ticket.

The fix to fetchSpecific is a bit ugly, but I don't see how to cleanly fix it without redefining its interface to return a more-fully-parsed result instead of the raw results of parseNestedParens.

I have to agree with exarkun that the IMAP spec requires the client to accept any responses at any time, and to take note of any that it needs to care about.  Unfortunately the twisted IMAP4 code in general is not structured to do this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [26390]) Branching to 'imap4-unsolicited-responses-1105'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [26391]) Apply twistedmail-imap-flags-anywhere-test.diff

refs [#10722](https://github.com/twisted/twisted/issues/10722)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [26392]) Coding standard updates etc

refs [#10722](https://github.com/twisted/twisted/issues/10722)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Should be somewhat more generally fixed now.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set owner to @exarkun|
|-|-|

  1. `IMAP4Client.__cbSelect` - you appear to have changed the `log.err` line such that `parts` is no longer in scope.  Pyflakes reports this as an error, and it has no test coverage.
  1. `pairs` may well be a lie, but I'd like to know what form of lie it is.  Please write a docstring for `_parseFetchPairs`.
  1. `XXX Used to strip here, still need to?` comments in `IMAP4Client.__cbSelect` - since these are all being modified... _do_ we still need to strip here?  What does this even mean?  File a new ticket?
  1. needs more tests:
    1. XXX in `IMAP4Client._store`
    1. comments at the bottom of `IMAP4ClientFetchTests` - I'm assuming these describe as-yet-unwritten tests, but I don't _really_ understand for what.  Enumerated here for your responding-to-feedback convenience:
      1. `test BODY[1.TEXT]`
      1. `test BODY[HEADER.FIELDS (SUBJECT)]`
      1. `test BODY.PEEK variations`
      1. `test error conditions`

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [26483]) Remove notes; I wrote those tests

refs [#10722](https://github.com/twisted/twisted/issues/10722)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [26484]) a docstring for _parseFetchPairs

refs [#10722](https://github.com/twisted/twisted/issues/10722)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [26486]) Get rid of XXXs about strip.  We don't need to anymore.

refs [#10722](https://github.com/twisted/twisted/issues/10722)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [26489]) Test some extra cases of the SELECT response parser (one of which turns out not to be real

refs [#10722](https://github.com/twisted/twisted/issues/10722)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun removed owner|
|-|-|

  1. Fixed in r26489.  The case wasn't actually reachable.  I added a test demonstrating this.
  2. Fixed in r26484.  
  3. Fixed in r26486.  Since the general parser is used now, there can't be random extra whitespace lying around unless the server tries really hard to put it there. I'm happy for such a case to not be handled by this code (if the server says _FETCH  foo_ instead of _FETCH foo_ that's one thing, probably some moron fat fingered it.  If the server says _FETCH " foo"_ though, then something fishy is going on).
  4. Fixed
    1. in r26485.  Added tests and fixed the implementation.
    2. in r26483.  I added these tests before the previous submission for review, I just forgot to delete the comments.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph set owner to @exarkun|
|-|-|

As we discussed on IRC, please fix the documentation for the `pairs` parameter.  Afterwards, go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

(In [26516]) Rename `pairs` something more accurate; adjust docstring and method to suit

refs [#10722](https://github.com/twisted/twisted/issues/10722)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set status to `closed`|
|-|-|

(In [26520]) Merge imap4-unsolicited-responses-1105 

Author: jhohm, exarkun
Reviewer: glyph
Fixes: [#10722](https://github.com/twisted/twisted/issues/10722)

Handle more unsolicited data received from the server instead of
printing it out and otherwise ignoring it.

In particular, now if unsolicited data is received on the same
line as solicited data, it is separated from the solicited data
and delivered to the appropriate callback.  Previously, this case
could cause the entire line to be ignored or it could cause the
unsolicited data to be delivered to the application callback with
the solicited data.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
